Runtime sysinfo check

runcom opened the pull request:
fix #15159 

tl;dr (more in commit messages)

sysinfo is initialized and stored in the daemon struct at daemon startup
runtime system settings changes aren't reflected in sysinfo daemon field
errors from libcontainer complaining about write to non existent cgroup path when it wasn't supposed to not even write there

now if a cgroup is umounted at runtime (memory for instance) and a container had memory-swappiness enabled and it's started again, the error from libcontainer is gone and in daemon logs you get

```
WARN[0108] Your kernel does not support cgroup memory limit: mountpoint for memory not found 
WARN[0108] Your kernel does not support memory swappiness capabilities, memory swappiness discarded. 
WARN[0108] IPv4 forwarding is disabled. Networking will not work 
WARN[0109] Your kernel does not support OOM notifications: There is no path for "memory" in state
```

**NOTE: warnings are not printed to stdout/stderr while starting a container (as it's happening for `run`) with wrong memory swappiness but only in daemon-logs.** This is because start return StatusNoContent so would not be good to break that behavior and print the warnings there and also this is more of a sysadmin thing who watches for wrong behaviors in daemon logs (please someone confirm this also)

I've made sysinfo checks at runtime and removed the field from the daemon struct
hostConfig is checked at container's start too

icecrime commented:
Do you think we should consider it for 1.8.0 (cc @calavera)?

runcom commented:
@icecrime yes I think, ppl don't usually umount cgroups at runtime or change sysctl settings but it may happen so better to have this check

I'll fix the panic/errors by tomorrow (too late here sorry)

hqhq commented:
Looks good :)

runcom commented:
@hqhq thanks, would be awesome if you give this patch a try :)

hqhq commented:
@runcom After some tests, I do find some problems:
1. It would be better if we can return warnings for start api, maybe this can leave to next release;
2. If you use `-m` instead of `--memory-swappiness` in this scenario, you'll get:
```
Error response from daemon: You should always set the Memory limit when using Memoryswap limit, see usage.
```
because memorySwap was set to double of memory on create, we need to handle this.
3. This problem also affect other cgroup subsystems, such as cpu and cpuset, but I think we'll need #14015 

But I confirm that problem for `memory-swappiness` has been resolved :smile:

runcom commented:
@hqhq perfect, thanks, I'll take care of your comments soon

runcom commented:
@hqhq I've fixed various things, please confirm everything is working as expected if you have any chance

calavera commented:
LGTM

runcom commented on daemon/daemon_unix.go:
```diff
@@ -131,20 +132,23 @@ func (daemon *Daemon) adaptContainerSettings(hostConfig *runconfig.HostConfig) {
 	}
+
+	// memory subsystem checks and adjustments
 	if hostConfig.Memory != 0 && hostConfig.Memory < 4194304 {
```
This check can be moved after the cgroup subsystem check, no point in showing this if memory cgroup isn't supported

hqhq commented on daemon/daemon_unix.go:
```diff
@@ -131,20 +132,23 @@ func (daemon *Daemon) adaptContainerSettings(hostConfig *runconfig.HostConfig) {
 		hostConfig.Memory = 0
 	}
-	if hostConfig.Memory > 0 && hostConfig.MemorySwap != -1 && !daemon.SystemConfig().SwapLimit {
+	if hostConfig.MemorySwap != -1 && !sysInfo.SwapLimit {
```
It would be a problem, after this change, we'll get:
```
$ docker run -ti ubuntu bash
WARNING: Your kernel does not support swap limit capabilities, memory limited without swap.
root@e22ad4765a95:/#
```
That warning was not expected because we don't want to limit swap in this command.

runcom replied:
Ok there's some kind of problem with default value there, I'll fix

hqhq commented on daemon/daemon_unix.go:
```diff
@@ -131,20 +132,23 @@ func (daemon *Daemon) adaptContainerSettings(hostConfig *runconfig.HostConfig) {
+	if hostConfig.Memory > 0 && !sysInfo.MemoryLimit {
 		warnings = append(warnings, "Your kernel does not support memory limit capabilities. Limitation discarded.")
 		logrus.Warnf("Your kernel does not support memory limit capabilities. Limitation discarded.")
 		hostConfig.Memory = 0
```
Maybe we can add `hostConfig.MemorySwap = -1` here.

runcom replied:
You mean also setting memory swap to - 1 when memory limit is disabled in kernel?

hqhq replied:
That's one solution for the problem I mentioned.
```
$ docker run --name ttt -m 300M busybox true
$ sudo umount /sys/fs/cgroup/memory
$ docker start ttt
```
In this scenario, after docker run, we'll have hostConfig.Memory = 300M and hostConfig.MemorySwap = 600M, then we umount memory cgroup and try to start container, `sysinfo.MemoryLimit` would be false, so we set `hostConfig.Memory = 0`, but then we'll get `hostConfig.Memory == 0 && hostConfig.MemorySwap > 0` and end up error.

So here when we set `hostConfig.Memory = 0` we can also set `hostConfig.MemorySwap = -1`, it'll fix it, and because if we don't limit memory, we can't limit memory swap either, so the logic is right.

runcom replied:
updated

runcom commented on daemon/daemon_unix.go:
```diff
@@ -155,7 +160,7 @@ func verifyPlatformContainerSettings(daemon *Daemon, hostConfig *runconfig.HostC
 		return warnings, fmt.Errorf("You should always set the Memory limit when using Memoryswap limit, see usage.")
 	}
-	if hostConfig.MemorySwappiness != nil && !daemon.SystemConfig().MemorySwappiness {
+	if hostConfig.MemorySwappiness != nil && !sysInfo.MemorySwappiness {
```
@hqhq now I get
```
./bundles/1.8.0-dev/binary/docker run -d debian:jessie true
WARNING: Your kernel does not support memory swappiness capabilities, memory swappiness discarded.
e8076e48660a6f7bded2ab67fb59a11c1878af79d78dff636957898e2803971e
```
but I suppose this is the correct behavior because Swappiness default value is -1

hqhq replied:
Emmm, then I think we should use:
`if hostConfig.MemorySwappiness != nil && *hostConfig.MemorySwappiness != -1 && !sysInfo.MemorySwappiness {`

-1 means don't set MemorySwappiness, let it inherit from parent, if we don't have Swappiness support, shouldn't show this warning.

runcom replied:
You're right, fixing

runcom committed:
Move sysinfo out of daemon struct

sysinfo struct was initialized at daemon startup to make sure
kernel configs such as device cgroup are present and error out if not.
The struct was embedded in daemon struct making impossible to detect
if some system config is changed at daemon runtime (i.e. someone
umount the memory cgroup). This leads to container's starts failure if
some config is changed at daemon runtime.
This patch moves sysinfo out of daemon and initilize and check it when
needed (daemon startup, containers creation, contaienrs startup for
now).

Signed-off-by: Antonio Murdaca <runcom@linux.com>

runcom committed:
Add hostConfig check before starting a container

It may happen that host system settings are changed while the daemon is running.
This will cause errors at libcontainer level when starting a container with a
particular hostConfig (e.g. hostConfig with memory swappiness but the memory
cgroup was umounted).
This patch adds an hostConfig check on container start to prevent the daemon
from even calling libcontainer with the wrong configuration as we're already
doing on container's creation).

Signed-off-by: Antonio Murdaca <runcom@linux.com>

calavera commented:
still LGTM

runcom commented:
@calavera I also feel that the other PRs about cgroups checks should go into the release if this is going in (otherwise seems only some checks are going into the release and others related aren't). I'll add the milestone to those one but feel free to remove it if it doesn't sound good

calavera commented:
let's evaluate them, I'm not completely sure which ones you're talking about

calavera commented:
Closing this in favor of https://github.com/docker/docker/pull/15381 that's already rebased, so we can release RC 3. Thanks @runcom :metal:
